src/tests/test_method_middleware.py

```diff
@@ -0,0 +1,324 @@
+            pass
+
+    dal = DataAccessLayer(data_manager=None, middleware=[SampleMiddleware], default_middleware=None)
+    assert isinstance(dal._reversed_middleware[0], SampleMiddleware)
+
+    dal = DataAccessLayer(data_manager=None, middleware=[SampleMiddleware()], default_middleware=None)
+    assert isinstance(dal._reversed_middleware[0], SampleMiddleware)
+
+
+def test_dal_middleware_custom_middleware():
+    """
+    Verify determining all middleware for the DAL to manage will
+    use our default middleware if none is specified, AND if any is
+    specified, our default middleware will be added as the last
+    middleware.
+
+    We also need to verify that these middlewares chain each other
+    in the order that we expect (reversed) and that they ultimately
+    use the default handler for dal methods.
+    """
+    # Verify default behavior first.
+    # Passing an empty DM here because we don't care about this
+    # for the purposes of the test, and it's not used.
+    dal = DataAccessLayer(data_manager=None)
+    assert len(dal._reversed_middleware) == 1
+    assert isinstance(dal._reversed_middleware[0], Callable)
+
+    # Verify custom default middleware
+    def example_custom_middleware(request, handler):
+        pass
+
+    dal = DataAccessLayer(data_manager=None, default_middleware=(example_custom_middleware,))
+    assert len(dal._reversed_middleware) == 1
+    assert dal._reversed_middleware[0] is example_custom_middleware
+
+    # Verify that even if you specify middleware, the default
+    # middleware gets appended to the end of the middleware stack.
+    # We test for this specifically because the default middleware
+    # needs to be run 'last'.
+
+    def foo_middleware(request, handler):
+        pass
+
+    def bar_middleware(request, handler):
+        pass
+
+    def baz_middleware(request, handler):
+        pass
+
+    middlewares = [foo_middleware, bar_middleware, baz_middleware]
+    dal = DataAccessLayer(data_manager=None, middleware=list(middlewares))
+
+    # The DAL handles making sure all of these are instantiated, but
+    # we have to manually do it for testing here.
+    middlewares.append(dal_method_resolver_middleware)
+    expected_middlewares = []
+    for m in reversed(middlewares):
+        expected_middlewares.append(m)
+
+    # Verify if the objects in dal._reversed_middleware are instances of the
+    # expected middleware types in the right order
+    for i in range(0, len(expected_middlewares)):
+        assert isinstance(dal._reversed_middleware[0], Callable)
+
+
+def test_dal_method_middleware():
+    """
+    Verify that when calling a method on a service off the DAL,
+    the method middlewares run, in the correct order and run
+    before the method itself is called.
+
+
+    We deliberately do not use the DalMethodResolverMiddleware here,
+    because we are testing middleware call chains, not functionality
+    of that class.
+    """
+    middleware_requests = []
```

Comment:
We normally only use these types of call hooks when we don't actually get the output. The problem is, no where are we actually testing that middleware can alter output. Below is a more black box test. It verifies input and output alteration without using hooks:

```python
def test_dal_method_middleware():
    """
    Verify that when calling a method on a service off the DAL,
    the method middlewares run, in the correct order and run
    before the method itself is called.


    We deliberately do not use the DalMethodResolverMiddleware here,
    because we are testing middleware call chains, not functionality
    of that class.
    """
    def test_method():
        return ['default-handler-call']

    class OuterMiddleware:
        def __call__(self, request, handler):
            result = ['outer-middleware-ingress']
            result.extend(handler(request))
            result.append('outer-middleware-egress')
            return result

    class SecondMiddleware:
        def __call__(self, request, handler):
            result = ['second-middleware-ingress']
            result.extend(handler(request))
            result.append('second-middleware-egress')
            return result

    # Verify custom default middleware
    class DefaultMiddleware:
        def __call__(self, request, handler):
            result = ['default-middleware-ingress']
            # We need to get past the DalMethodResolverMiddleware
            # which sets the `dal_method` attribute to
            # the method that has to be called
            request.dal_method = test_method
            result.extend(handler(request))
            result.append('default-middleware-egress')
            return result

    dm = DataManager()

    dal = DataAccessLayer(data_manager=dm, middleware=[OuterMiddleware(), SecondMiddleware()],
        default_middleware=(DefaultMiddleware(),))

    with dm.context():
        result = dal.fake.service.method()

    assert result == [
        'outer-middleware-ingress',
        'second-middleware-ingress',
        'default-middleware-ingress',
        'default-handler-call',
        'default-middleware-egress',
        'second-middleware-egress',
        'outer-middleware-egress',
    ]
```